libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -682,28 +682,46 @@ string YulUtilFunctions::overflowCheckedIntMulFunction(IntegerType const& _type)
 			function <functionName>(x, y) -> product {
 				x := <cleanupFunction>(x)
 				y := <cleanupFunction>(y)
+				product := mul(x, y)
 				<?signed>
-					// overflow, if x > 0, y > 0 and x > (maxValue / y)
-					if and(and(sgt(x, 0), sgt(y, 0)), gt(x, div(<maxValue>, y))) { <panic>() }
-					// underflow, if x > 0, y < 0 and y < (minValue / x)
-					if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(<minValue>, x))) { <panic>() }
-					// underflow, if x < 0, y > 0 and x < (minValue / y)
-					if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { <panic>() }
-					// overflow, if x < 0, y < 0 and x < (maxValue / y)
-					if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { <panic>() }
+					<?256bit>
+						// special case
+						if and(eq(x, sub(0, 1)), eq(y, <minValue>)) { <panic>() }
+						if and(
+							iszero(iszero(x)),
+							iszero(eq(y, sdiv(product, x)))
+						) { <panic>() }
+					<!256bit>
+						// overflow if product has positive sign and y != (p & bitMask) / x
+						// 			OR product has negative sign and y != (p | !bitMask) / x
+						if and(
+							iszero(iszero(x)),
+							or(
+								and(
+									iszero(and(product, <signMask>)),
+									iszero(eq(y, sdiv(and(product, <bitMask>),x)))
+								),
+								and(
+									eq(and(product, <signMask>), <signMask>),
```

Comment:
Can we add a single `iszero` here and move all the other `iszero`s out using demorgan?